{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/gilsonauerswald/Bioinformatic_Projects/blob/main/Python_C1_03_Reading_DNA_using_Python.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Welcome Back to Python Course 1!**\n",
        "\n",
        "In the previous lesson, we covered data visualization using various python packages like Matplotlib and seaborn for the better understanding of our data. If you haven't completed these topics, please refer to L02 - Data visualization in Python, before practicing the codes for this lesson."
      ],
      "metadata": {
        "id": "CUOl1CXyY9ij"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Let's Recap!**\n",
        "\n",
        "Now that we've covered essential concepts in [L03 - Reading DNA using Python](https://learn.omicslogic.com/Python/python-course-1-getting-started-with-bioinformatics/lesson/03-reading-dna-using-python), let's take a moment to reinforce what we've learned so far. We'll start with a brief recap of the codes and techniques we've explored in the lesson."
      ],
      "metadata": {
        "id": "eevZuVgmiUNK"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Lesson Code Breakdown**"
      ],
      "metadata": {
        "id": "Mr1f6Vl3b8Lu"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **1. Counting DNA length**"
      ],
      "metadata": {
        "id": "oEBr8wg0ZAnz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#define our functions\n",
        "def print_sequence_length():\n",
        "  #store DNA strand as a string in the variable seq\n",
        "  seq=\"AGTGTCCCTG\"\n",
        "  #returns the sequence length\n",
        "  return len(seq)\n",
        "\n",
        "\n",
        "#run our function\n",
        "print_sequence_length()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "NxjZA4FMZBlb",
        "outputId": "149f7877-3a1f-4566-9052-5afaabbffce0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "10"
            ]
          },
          "metadata": {},
          "execution_count": 1
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "The given code defines a function named print_sequence_length without any input parameters. Inside the function, a DNA strand is stored as a string in the variable seq. The function uses the len function to calculate the length of the DNA sequence stored in seq and returns the result.\n",
        "\n",
        "After defining the function, the code snippet calls the function print_sequence_length using the print function. This will execute the function and print the returned value, which is the length of the DNA sequence \"AGTGTCCCTG\"."
      ],
      "metadata": {
        "id": "QCNdhtDxioeZ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **2. Calculate nucleotide frequency**"
      ],
      "metadata": {
        "id": "aVi1VL5qbJPZ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def calculate_nucleotide_frequency(dna_sequence):\n",
        "    # Calculate the frequency of each nucleotide\n",
        "    nucleotide_count = {'A': 0, 'T': 0, 'C': 0, 'G': 0}\n",
        "\n",
        "    for nucleotide in dna_sequence:\n",
        "        if nucleotide in nucleotide_count:\n",
        "            nucleotide_count[nucleotide] += 1\n",
        "\n",
        "    return nucleotide_count\n",
        "\n",
        "\n",
        "\n",
        "# Example DNA sequence\n",
        "dna_sequence = \"ATGCTTCAGAAAGGTCTTACG\"\n",
        "\n",
        "# Compute frequencies and GC content\n",
        "nucleotide_frequencies = calculate_nucleotide_frequency(dna_sequence)\n",
        "\n",
        "\n",
        "# Display the results\n",
        "print(\"Nucleotide Frequencies:\", nucleotide_frequencies)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "njeS8uBgbTh1",
        "outputId": "1cf6cd3d-90aa-457f-9e79-bed5cc2612b7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Nucleotide Frequencies: {'A': 6, 'T': 6, 'C': 4, 'G': 5}\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **3. Calculate GC content**"
      ],
      "metadata": {
        "id": "51nIAFrubm6D"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def calculate_gc_content(dna_sequence):\n",
        "    # Calculate GC content\n",
        "    gc_count = sum(dna_sequence.count(nucleotide) for nucleotide in 'GC')\n",
        "    gc_content = (gc_count / len(dna_sequence)) * 100\n",
        "    return gc_content\n",
        "\n",
        "# Example DNA sequence\n",
        "dna_sequence = \"ATGCTTCAGAAAGGTCTTACG\"\n",
        "\n",
        "gc_content = calculate_gc_content(dna_sequence)\n",
        "\n",
        "print(\"GC Content (%):\", gc_content)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "CTRSTsNDbW0c",
        "outputId": "12050e6e-4602-42a7-dc29-08f25eb6f212"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "GC Content (%): 42.857142857142854\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **4. Function to download and read a FASTA file**"
      ],
      "metadata": {
        "id": "AZ-wmlb3ZKKm"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**FASTA sequences**\n",
        "\n",
        "**>OmicsLogic_1218**\n",
        "\n",
        "CCTGCGGAAGATCGGCGTCGGTCTAGCCAGAACCGTTTCTCTGAGGCTTCCGGCCTTCCC TCCCACTAATAATTCTGAGC\n",
        "\n",
        "**>OmicsLogic_2730**\n",
        "\n",
        "CCATCGGTAGCGCTTCCTTCGGCTAATTAAGTCCCTATCCAGGCGCTCCGACGAAGGTCT ATATCCATTTGTCAGCAGACACTC\n",
        "\n",
        "**>OmicsLogic_4280**\n",
        "\n",
        "CCACCCTCGTGGTATGGCTAGGCATTTTCGAACCGGATAACGCTTCAGACAAGCCCGGAC TGGGAACCTGCGCGCAGTAGGTAGAAT"
      ],
      "metadata": {
        "id": "ld6U88hZZCuc"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "The given text is a collection of DNA sequences in FASTA format. Each sequence is represented by a header line starting with the \">\" symbol, followed by the sequence itself.\n",
        "\n",
        "Each sequence is a string of characters representing the DNA bases (A, T, G, C). The sequences can be used for various bioinformatics analyses or processing."
      ],
      "metadata": {
        "id": "5-nL0HHGi20Q"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#imports a library included in Python to give extra functions\n",
        "import requests\n",
        "from collections import Counter\n",
        "\n",
        "# Function to download and parse a FASTA file from a URL, returning a dictionary {'label':'seq'}.\n",
        "def download_and_parse_fasta(url):\n",
        "    try:\n",
        "        response = requests.get(url)\n",
        "        response.raise_for_status()  # Will raise an HTTPError if the HTTP request returned an unsuccessful status code\n",
        "\n",
        "        FASTA_dict = {}\n",
        "        FASTA_label = \"\"\n",
        "        for line in response.text.split('\\n'):\n",
        "            line = line.rstrip()  # Removes spaces and newline chars each line\n",
        "            if line.startswith(\">\"):  # Catches sequence label\n",
        "                FASTA_label = line[1:]  # Set the label as 1st char and after\n",
        "                FASTA_dict[FASTA_label] = \"\"  # Set sequence label in dict with empty string\n",
        "            else:  # Must be sequence\n",
        "                FASTA_dict[FASTA_label] += line\n",
        "        return FASTA_dict\n",
        "    except requests.RequestException as e:\n",
        "        print(f\"Error downloading file: {e}\")\n",
        "        return None\n"
      ],
      "metadata": {
        "id": "ArMsKblFkGUF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "The code snippet imports the Counter class from the collections module and defines the download_and_parse_fasta function. This function reads a FASTA file, parses its sequences into a dictionary using labels as keys, and returns the dictionary. However, the code does not show the complete implementation, including base occurrence counting and result printing."
      ],
      "metadata": {
        "id": "6KeFgv23nQTR"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **5. Function to count base occurrence in a FASTA file**"
      ],
      "metadata": {
        "id": "BKzPvMy9ZNfG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Function to count base occurrence and print results\n",
        "def count_bases_from_url(url, sequence_label):\n",
        "    # Run our download FASTA function and set returned FASTA_dict to FASTA variable\n",
        "    FASTA = download_and_parse_fasta(url)\n",
        "    if FASTA is None:\n",
        "        return None\n",
        "\n",
        "    # Select the sequence we want to look at from the dict\n",
        "    sequence = FASTA.get(sequence_label)\n",
        "    if sequence is None:\n",
        "        print(f\"Sequence label '{sequence_label}' not found in the FASTA file.\")\n",
        "        return None\n",
        "\n",
        "    # Counts and returns the bases in the genome as a Counter object\n",
        "    return Counter(sequence)\n",
        "\n",
        "# Example usage (you'll need to replace 'your_fasta_file_url' with the actual URL and the sequence label)\n",
        "url = 'https://raw.githubusercontent.com/pine-bio-support/DNA/master/seq1.fa'\n",
        "sequence_label = 'seq1'\n",
        "print(str(count_bases_from_url(url, sequence_label)))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XHLg7Y16pOGd",
        "outputId": "03d7acc1-d9ba-4326-9b2e-47da1dc70860"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Counter({'A': 4, 'T': 3, 'G': 3, 'C': 1})\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "The code snippet imports the Counter class from the collections module and defines the count_bases function. This function uses the download_and_parse_fasta function to retrieve a specific sequence, from a FASTA file, then employs the Counter class to count the occurrences of each base in the sequence. The result is converted to a string for output."
      ],
      "metadata": {
        "id": "By5nEdJaoYpo"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Practice Codes: Try it yourself**"
      ],
      "metadata": {
        "id": "QNE2cWy4Zb8F"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **Assignment 1:**\n",
        "In this assignment, you will build upon your knowledge of reading DNA using Python. The provided code is incomplete, with missing syntax and portions. Your task is to complete the code to successfully count the DNA length.To complete this, use the following sequence \"AGTGTCCCTGTGTC\"."
      ],
      "metadata": {
        "id": "cfm_RgHLZeGM"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#define our functions\n",
        "def print_sequence_length(:\n",
        "  seq = \"\" #store DNA strand as a string in the variable seq\n",
        "  return len(seq) #returns the sequence length\n",
        "\n",
        "print_sequence_length() #run our function"
      ],
      "metadata": {
        "id": "kin_kstzZfOp",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 140
        },
        "outputId": "87cbc9f9-84d7-42a7-a792-d52958720fcf"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "SyntaxError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-5-cd892a4bcaac>\"\u001b[0;36m, line \u001b[0;32m2\u001b[0m\n\u001b[0;31m    def print_sequence_length(:\u001b[0m\n\u001b[0m                              ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **Assignment 2:**\n",
        "In this assignment, you will build upon your knowledge of reading DNA using Python. The provided code is incomplete, with missing syntax and portions. Your task is to complete the code to successfully calculating nucleotide frequency. To complete this, use the following sequence \"AGTGTCCCTGTGTC\"."
      ],
      "metadata": {
        "id": "v-tmJJJo55AF"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def calculate_nucleotide_frequency(:\n",
        "    # Calculate the frequency of each nucleotide\n",
        "    nucleotide_count = {'A': 0, 'T': 0, 'C': 0, 'G': 0}\n",
        "\n",
        "    for nucleotide in dna_sequence:\n",
        "        if nucleotide in nucleotide_count:\n",
        "            nucleotide_count[nucleotide] += 1\n",
        "\n",
        "    return nucleotide_count\n",
        "\n",
        "\n",
        "\n",
        "# Example DNA sequence\n",
        "dna_sequence = \"\"\n",
        "\n",
        "# Compute frequencies and GC content\n",
        "nucleotide_frequencies = calculate_nucleotide_frequency(dna_sequence)\n",
        "\n",
        "\n",
        "# Display the results\n",
        "print(\"Nucleotide Frequencies:\", nucleotide_frequencies)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "AH7_iqZ76RZN",
        "outputId": "bbb0963b-4a83-4a18-8657-092599a5815a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Nucleotide Frequencies: {'A': 0, 'T': 0, 'C': 0, 'G': 0}\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **Assignment 3:**\n",
        "In this assignment, you will build upon your knowledge of reading DNA using Python. The provided code is incomplete, with missing syntax and portions. Your task is to complete the code to successfully calculating GC content. To complete this, use the following sequence \"AGTGTCCCTGTGTC\"."
      ],
      "metadata": {
        "id": "ysk2F9TB8l-t"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def calculate_gc_content(:\n",
        "    # Calculate GC content\n",
        "    gc_count = sum(dna_sequence.count(nucleotide) for nucleotide in 'GC')\n",
        "    gc_content = (gc_count / len(dna_sequence)) * 100\n",
        "    return gc_content\n",
        "\n",
        "# Example DNA sequence\n",
        "dna_sequence = \"\"\n",
        "\n",
        "gc_content = calculate_gc_content(dna_sequence)\n",
        "\n",
        "print(\"GC Content (%):\", gc_content)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 140
        },
        "id": "y5QNctQv8w8Z",
        "outputId": "dbedaeab-65d7-445e-9574-612f7840e095"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "SyntaxError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-11-acd22d665bd0>\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    def calculate_gc_content(:\u001b[0m\n\u001b[0m                             ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **Assignment 4:**\n",
        "In this assignment, you will build upon your knowledge of reading DNA using Python. The provided code is incomplete, with missing syntax and portions. Your task is to complete the code to successfully import a FASTA file, read it and then count the base occurrence. To complete this, use the following fasta file: https://raw.githubusercontent.com/pine-bio-support/DNA/master/seq1.fa, download it and then upload it in colab to use."
      ],
      "metadata": {
        "id": "Q35YTBjO9hFb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#imports a library included in Python to give extra functions\n",
        "from collections import Counter\n",
        "\n",
        "\"\"\"function takes in a filepath then opens & parses the\n",
        "fasta file returning a dictionary {'label':'seq'}.\"\"\"\n",
        "def open_and_parse_fasta(filepath\n",
        "    FASTA_file = open(f\"{filepath}\", \"r\")\n",
        "    FASTA_dict = {}\n",
        "    FASTA_label = \"\"\n",
        "    for line in FASTA_file:\n",
        "        line = line.rstrip() #removes spaces and newline chars each line\n",
        "        if line.startswith(\">\"): #catches sequence label\n",
        "            FASTA_label = line[1:] #set the label as 1st char and after\n",
        "            #set sequence label in dict with empty string\n",
        "            FASTA_dict[FASTA_label] = \"\"\n",
        "        else: #must be sequence\n",
        "            #add sequence to end of dict entry\n",
        "            FASTA_dict[FASTA_label] += line\n",
        "    return FASTA_dict\n",
        "\n",
        "#function to count base occurence and print results\n",
        "def count_bases():\n",
        "  #run our open FASTA function and set returned FASTA_dict to FASTA variable\n",
        "  FASTA = open_and_parse_fasta(\"file path\")\n",
        "  #select the sequence we want to look at from the dict\n",
        "  carsonella_ruddi = FASTA.get('Sequence label')\n",
        "  #counts and returns the bases in the genome as a counter object\n",
        "  return Counter(carsonella_ruddi)\n",
        "\n",
        "#run our function and print it as a string\n",
        "str(count_bases())"
      ],
      "metadata": {
        "id": "itFQknQiZhog",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 140
        },
        "outputId": "4a6c730a-f414-4610-f635-1ec3327994f4"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "SyntaxError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-12-661a71466916>\"\u001b[0;36m, line \u001b[0;32m6\u001b[0m\n\u001b[0;31m    def open_and_parse_fasta(filepath\u001b[0m\n\u001b[0m                            ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m '(' was never closed\n"
          ]
        }
      ]
    }
  ]
}