{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/gilsonauerswald/Bioinformatic_Projects/blob/main/R_06_Advanced_analysis_of_VCF_files_Quality_control%2C_Filter_and_visualize_the_genomic_variants.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Install Necessary Packages**"
      ],
      "metadata": {
        "id": "QdoWloNxCRry"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Install packages\n",
        "install.packages(\"vcfR\")\n",
        "install.packages(\"reshape2\")\n",
        "#Load packages\n",
        "library(\"vcfR\")\n",
        "library(\"reshape2\")\n",
        "library('ggplot2')"
      ],
      "metadata": {
        "id": "NbHSNZOvCUdR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "fRJ15rbH3png"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Load the Sample Data**"
      ],
      "metadata": {
        "id": "gtQAQgPGCWNE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Load data\n",
        "sample_vcf_tp53 <- read.vcfR('https://raw.githubusercontent.com/pine-bio-support/Merge-VCF-files/main/aneuploid_samples_freebayes_tp53.vcf')"
      ],
      "metadata": {
        "id": "rqnfnsVPCYFH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "wvjqFSht3qGu"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#show the vcfR object\n",
        "sample_vcf_tp53"
      ],
      "metadata": {
        "id": "yQmxH04bCZwG"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "2pmfM_XC3qnh"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**The vcfR Object**"
      ],
      "metadata": {
        "id": "kk9kLtinCbEb"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Display first few lines of each slots\n",
        "head(sample_vcf_tp53)"
      ],
      "metadata": {
        "id": "O9cicKH5CcTa"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "T3Yk-jCc3rS_"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Querying the Meta Data**"
      ],
      "metadata": {
        "id": "MzOsYU8kCdpF"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Querying meta data for element DP\n",
        "queryMETA(sample_vcf_tp53, element = 'FORMAT.+DP')"
      ],
      "metadata": {
        "id": "MFzZssDhCsEP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "FgURNY_83sB5"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Extract depth (DP)**"
      ],
      "metadata": {
        "id": "GpxijdsVCxmY"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Genotype GT data of the sample\n",
        "sample_vcf_tp53@gt[1:5,1:5]"
      ],
      "metadata": {
        "id": "21BjW4pHCufl"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "r8arcGb43spG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Extract Depth from the genotype portion\n",
        "dp <- extract.gt(sample_vcf_tp53, element = \"DP\", as.numeric=TRUE)\n",
        "head(dp)"
      ],
      "metadata": {
        "id": "VrPfGAaoDFNP"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "ojPY1kDG3tJr"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Depth Plot**"
      ],
      "metadata": {
        "id": "vzSdDIUyDHme"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Set the margin, width and height for the plot\n",
        "par(mar=c(12,4,4,2))\n",
        "#Box plot the depth matrix for selected columns\n",
        "boxplot(dp, col=2:ncol(dp), las=3)\n",
        "#Give a title\n",
        "title(ylab = \"Depth (DP)\")"
      ],
      "metadata": {
        "id": "qhkbUMA4DJIK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "HJA5y6Pk3tm-"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Log Transformation**"
      ],
      "metadata": {
        "id": "O6qlTlprDK6h"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Set the margin, width and height for the plot\n",
        "par(mar=c(12,4,4,2))\n",
        "#Box plot the depth matrix for selected columns in log scale\n",
        "boxplot(dp, col=2:ncol(dp), las=3, log='y')\n",
        "title(ylab = \"Depth (DP)\")"
      ],
      "metadata": {
        "id": "cd9VX-0LDMZi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "WyIlYCCT3uEM"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Filtering on sequence depth.**"
      ],
      "metadata": {
        "id": "i_wGtanbDNhv"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dp_filt <- dp\n",
        "sums <- apply(dp_filt, MARGIN=2, quantile, probs=c(0.05, 0.95), na.rm=TRUE)\n",
        "dp2 <- sweep(dp_filt, MARGIN=2, FUN = \"-\", sums[1,])\n",
        "dp_filt[dp2 < 0] <- NA\n",
        "dp2 <- sweep(dp_filt, MARGIN=2, FUN = \"-\", sums[2,])\n",
        "dp_filt[dp2 > 0] <- NA\n",
        "dp_filt[dp_filt < 4] <- NA"
      ],
      "metadata": {
        "id": "gS8sFrZIDO5A"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "Ql1-KQli3ukK"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Set the margin, width and height for the plot\n",
        "par(mar=c(12,4,4,2))\n",
        "#Box plot the depth matrix for selected columns in log scale\n",
        "boxplot(dp_filt, col=2:ncol(dp), las=3, log='y')\n",
        "title(ylab = \"Depth (DP)\")"
      ],
      "metadata": {
        "id": "hwW19SBYDQEC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "bUGqUKx13wCn"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# show the vcfR object\n",
        "sample_vcf_tp53"
      ],
      "metadata": {
        "id": "qM8NgAdhDRNg"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "JNTpWTui3whF"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Employ filtering on a duplicated vcfR object based on the filtered depth matrix.\n",
        "sample_vcf_tp53_filt <- sample_vcf_tp53\n",
        "is.na( sample_vcf_tp53_filt@gt[,-1][ is.na(dp_filt) ] ) <- TRUE\n",
        "sample_vcf_tp53_filt"
      ],
      "metadata": {
        "id": "1idqwxc4DSTq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "eJTUq_MX3xE0"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# **Missing data**"
      ],
      "metadata": {
        "id": "xAADBlE3DUCE"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **Quantifying missing values across all samples**"
      ],
      "metadata": {
        "id": "PeLt4d6QDWE9"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "sample_vcf_tp53"
      ],
      "metadata": {
        "id": "9x38heu9DXuA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "72sm9CUs3xuc"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **Quantifying missing values in one sample**"
      ],
      "metadata": {
        "id": "v1HskPmYDY7B"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "Quantifying missing values in one sample"
      ],
      "metadata": {
        "id": "iGILzgcRDanO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "jYn_A8953ya8"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## **Quantifying missing values across all samples**"
      ],
      "metadata": {
        "id": "MEHkwjd5DbtX"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Estimate the number of missing values across different samples\n",
        "myMiss <- apply(dp, MARGIN = 2, function(x){ sum(is.na(x)) })\n",
        "#Estimate the proportion of missing values by number of variants\n",
        "myMiss <- myMiss/nrow(sample_vcf_tp53)\n",
        "\n",
        "\n",
        "#Use an extended colour pallete\n",
        "library(RColorBrewer)\n",
        "palette(brewer.pal(n=12, name = 'Set3'))\n",
        "\n",
        "\n",
        "#Set the margin, width and height of the plot\n",
        "par(mar = c(12,4,4,2))\n",
        "\n",
        "\n",
        "#Barplot for the missing values\n",
        "barplot(myMiss, las = 2, col = 1:12)\n",
        "title(ylab = \"Missing Variants (%)\")"
      ],
      "metadata": {
        "id": "AbSu-qgFDddA"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "k74lRAgN3zCO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Estimate the number of missing values across different variants\n",
        "myMiss <- apply(dp, MARGIN = 1, function(x){ sum(is.na(x)) })\n",
        "#Estimate the proportion of missing values by number of samples\n",
        "myMiss <- myMiss/ncol(sample_vcf_tp53@gt[,-1])\n",
        "\n",
        "\n",
        "#Histogram of missing values across variants\n",
        "hist(myMiss, col = \"#8DD3C7\", xlab = \"Missing Variants (%)\",  breaks = 5)"
      ],
      "metadata": {
        "id": "FCbtWe6lDer_"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}